net/sched/sch_gkprio.c

```diff
@@ -0,0 +1,298 @@
+			   struct sk_buff **to_free)
+{
+	struct qdisc_skb_head *qdisc;
+	struct qdisc_skb_head *lp_qdisc;
+	struct sk_buff *to_drop;
+	int wlen, ret;
+	u16 prio, lp;
+	struct gkprio_sched_data *q;
+
+	/* Obtain the priority of @skb. */
+	wlen = skb_network_offset(skb);
+	switch (tc_skb_protocol(skb)) {
+	case htons(ETH_P_IP):
+		wlen += sizeof(struct iphdr);
+		if (!pskb_may_pull(skb, wlen))
+			goto drop;
+		prio = ipv4_get_dsfield(ip_hdr(skb)) >> 2;
+		break;
+
+	case htons(ETH_P_IPV6):
+		wlen += sizeof(struct ipv6hdr);
+		if (!pskb_may_pull(skb, wlen))
+			goto drop;
+		prio = ipv6_get_dsfield(ipv6_hdr(skb)) >> 2;
+		break;
+
+	default:
+		prio = q->noip_dfltp;
+		break;
+	}
+
+	q = qdisc_priv(sch);
+	qdisc = &q->qdiscs[prio];
+
+	if (sch->q.qlen < q->max_limit) {
+		ret = __qdisc_enqueue_tail(skb, sch, qdisc);
+		if (ret < 0)
+			return ret;
+
+		/* Check to update highest and lowest priorities. */
+		if (prio > q->highest_prio)
+			q->highest_prio = prio;
+
+		if (prio < q->lowest_prio)
+			q->lowest_prio = prio;
+
+		sch->q.qlen++;
+		return ret;
+	}
+
+	/* If this packet is lowest priority, drop it. */
+	lp = q->lowest_prio;
+	if (prio <= lp)
+		return qdisc_drop(skb, sch, to_free);
+
+	/* Drop the packet at the tail of lowest priority qdisc. */
+	lp_qdisc = &q->qdiscs[lp];
+	to_drop = lp_qdisc->tail;
+
+        if (likely(to_drop)) {
+                lp_qdisc->tail = to_drop->prev;
+                lp_qdisc->qlen--;
+                if (!lp_qdisc->tail)
+                        lp_qdisc->head = NULL;
+                to_drop->prev = NULL;
+		qdisc_qstats_backlog_dec(sch, to_drop);
+		 __qdisc_drop(to_drop, to_free);
+		qdisc_qstats_drop(sch);
+        }	
+	else goto drop;
+
+	ret = __qdisc_enqueue_tail(skb, sch, qdisc);
+	if (ret < 0)
+		return ret;
+
+	/* Check to update highest and lowest priorities. */
+	if (lp_qdisc->qlen == 0)
+		q->lowest_prio = calc_new_low_prio(q);
+
+	if (prio > q->highest_prio)
+		q->highest_prio = prio;
+
+	return ret;
+drop:
+	qdisc_drop(skb, sch, to_free);
+	return NET_XMIT_SUCCESS | __NET_XMIT_BYPASS;
+}
+
+static struct sk_buff *gkprio_dequeue(struct Qdisc *sch)
+{
+	struct gkprio_sched_data *q = qdisc_priv(sch);
+	struct sk_buff *skb;
+	struct qdisc_skb_head *hpq = &q->qdiscs[q->highest_prio];
+
+	skb = __qdisc_dequeue_head(hpq);
+	if (unlikely(!skb))
+		return NULL;
+
+	sch->q.qlen--;
+	qdisc_qstats_backlog_dec(sch, skb);
+	qdisc_bstats_update(sch, skb);
+
+	/* Update highest priority field. */
+	if (hpq->qlen == 0)
+		q->highest_prio = calc_new_high_prio(q);
+
+	return skb;
+}
+
+static int gkprio_init(struct Qdisc *sch, struct nlattr *opt)
+{
+	int prio;
+	struct gkprio_sched_data *q = qdisc_priv(sch);
+
+	if (!opt) {
+		q->max_limit = qdisc_dev(sch)->tx_queue_len;
+		q->noip_dfltp = 0;
+	} else {
+		struct tc_gkprio_qopt *ctl = nla_data(opt);
+
+		if (nla_len(opt) < sizeof(*ctl))
+			return -EINVAL;
+
+		q->max_limit = ctl->limit;
+		q->noip_dfltp = ctl->dfltp;
```

Comment:
1. The code must check that `ctl->limit` and `ctl->dfltp` have valid numbers.

2. Both parameters must be optional, that is, any of them can be missing. You can use the value with all bits on (i.e. -1) to mark a parameter is missing.

Reply:
I've left checking for `ctl->dfltp` to be done in userspace since we know the possible priorities that are allowed, and `ctl->limit` checking is automatically done, hope this is ok.

Reply:
Kernel code cannot rely on usespace code to sanitize its values. The reason for this rule is that someone can write another userspace application that tries to trick the kernel sending unsanitized values. All kernel code must make sure that whatever it gets from userland is safe before using the information.

This safety verification may lead to duplicate sanitation code: one in the kernel and another in userland applications. But there's a difference here. The code in the kernel is required and essential to keep the safety of the system, whereas the code in the userland application is there mostly to provide easier to understand error messages.